src/engine/BMDieWildcard.php

```diff
@@ -9,4 +9,294 @@
+        } elseif (54 === $this->currentCardId) {
+            $textStr = 'Jkr (black)';
+            $htmlStr = '<span class="suit_black">Jkr</span>';
+        } else {
+            if ($this->currentCardId >= 40) {
+                $suit = 'S';
+                $htmlSuit = '<span class="suit_black">&spades;</span>';
+            } elseif ($this->currentCardId >= 27) {
+                $suit = 'H';
+                $htmlSuit = '<span class="suit_red">&hearts;</span>';
+            } elseif ($this->currentCardId >= 14) {
+                $suit = 'D';
+                $htmlSuit = '<span class="suit_red">&diams;</span>';
+            } else {
+                $suit = 'C';
+                $htmlSuit = '<span class="suit_black">&clubs;</span>';
+            }
+
+            if (1 === $this->currentCardId % 13) {
+                $value = 'A';
+            } elseif (11 === $this->currentCardId % 13) {
+                $value = 'J';
+            } elseif (12 === $this->currentCardId % 13) {
+                $value = 'Q';
+            } elseif (0 === $this->currentCardId % 13) {
+                $value = 'K';
+            } else {
+                $value = $this->currentCardId % 13;
+            }
+
+            $htmlStr = $value . $htmlSuit;
+            $textStr = $value . $suit;
+        }
+
+        if ($isHtml) {
+            return $htmlStr;
+        } else {
+            return $textStr;
+        }
+    }
+
+    /**
+     * Set number of sides of the die, and add die skills
+     *
+     * @param int $sides
+     * @param array $skills
+     */
+    public function init($sides, array $skills = NULL) {
+        $this->currentCardId = 0;
+        $this->min = 1;
+        $this->max = 20;
+
+        $this->add_multiple_skills($skills);
+    }
+
+    /**
+     * Create a BMDieWildcard, then add skills to the die.
+     *
+     * @param array $skills
+     * @return BMDieWildcard
+     */
+    public static function create($size, array $skills = NULL) {
+        $die = new BMDieWildcard;
+
+        $die->init($size, $skills);
+
+        return $die;
+    }
+
+    /**
+     * Select a new value
+     */
+    protected function select_new_value() {
+        if (!($this->ownerObject instanceof BMGame)) {
+            return;
+        }
+
+        $game = $this->ownerObject;
+        $playerArray = $game->playerArray;
+        $player = $game->playerArray[$this->playerIdx];
+
+        if (count($player->cardsDrawn) >= 54) {
+            $player->cardsDrawn = array();
+        }
+
+        $cardsDrawn = $player->cardsDrawn;
+
+        while (TRUE) {
+            $newCardId = bm_rand(1, 54);
+            $cardsDrawn[$newCardId] = TRUE;
+
+            if (($this->max < 20) &&
+                ($this->get_value_from_id($newCardId) > $this->max)) {
+                continue;
+            }
+
+            if (!array_key_exists($newCardId, $player->cardsDrawn)) {
+                break;
+            }
+        }
```

Comment:
I haven't been paying attention, but this implementation seems really needlessly wasteful --- why on earth don't you just make an array containing the numbers in your range, shuffle it, and iterate over it until you find a good one?  I don't have the stats top-of-mind to prove this, but it seems to me that if `$this->max` is small, it could take significantly more than 54 tries to find the valid number.  So why wouldn't you generate and shuffle the deck once, so that you're guaranteed that the loop with terminate in a known number of tries?